crates/driver/src/domain/competition/auction.rs

```diff
@@ -55,31 +92,111 @@ impl Auction {
                 class,
                 // If the orders are of the same kind, then sort by likelihood of fulfillment
                 // based on token prices.
-                order.likelihood(&tokens),
+                order.likelihood(&self.tokens),
             ))
         });
 
-        // TODO Filter out orders based on user balance
+        // Fetch balances of each token for each trader.
+        // Has to be separate closure due to compiler bug.
+        let f = |order: &competition::Order| -> (order::Trader, eth::TokenAddress) {
+            (order.trader(), order.sell.token)
+        };
+        let tokens_by_trader = self.orders.iter().map(f).unique();
+        let mut balances: HashMap<
+            (order::Trader, eth::TokenAddress),
+            Result<eth::TokenAmount, crate::infra::blockchain::Error>,
+        > = join_all(tokens_by_trader.map(|(trader, token)| async move {
+            let balance = eth.balance_of(trader.into(), token).await;
```

Comment:
We would need to pull in the `BalanceFetching` implementation where we simulate with hooks here to not break the current hooks implementation.